test/e2e/framework/initialize.go

```diff
@@ -0,0 +1,171 @@
+	ExpectNoError(WaitForAllNodesSchedulable(c, TestContext.NodeSchedulableTimeout))
+
+	// If NumNodes is not specified then auto-detect how many are scheduleable and not tainted
+	if TestContext.CloudConfig.NumNodes == DefaultNumNodes {
+		TestContext.CloudConfig.NumNodes = len(GetReadySchedulableNodesOrDie(c).Items)
+	}
+
+	// Ensure all pods are running and ready before starting tests (otherwise,
+	// cluster infrastructure pods that are being pulled or started can block
+	// test pods from running, and tests that ensure all pods are running and
+	// ready will fail).
+	podStartupTimeout := TestContext.SystemPodsStartupTimeout
+	// TODO: In large clusters, we often observe a non-starting pods due to
+	// #41007. To avoid those pods preventing the whole test runs (and just
+	// wasting the whole run), we allow for some not-ready pods (with the
+	// number equal to the number of allowed not-ready nodes).
+	if err := e2epod.WaitForPodsRunningReady(c, metav1.NamespaceSystem, int32(TestContext.MinStartupPods), int32(TestContext.AllowedNotReadyNodes), podStartupTimeout, map[string]string{}); err != nil {
+		DumpAllNamespaceInfo(c, metav1.NamespaceSystem)
+		LogFailedContainers(c, metav1.NamespaceSystem, e2elog.Logf)
+		runKubernetesServiceTestContainer(c, metav1.NamespaceDefault)
+		e2elog.Failf("Error waiting for all pods to be running and ready: %v", err)
+	}
+
+	if err := WaitForDaemonSets(c, metav1.NamespaceSystem, int32(TestContext.AllowedNotReadyNodes), TestContext.SystemDaemonsetStartupTimeout); err != nil {
+		e2elog.Logf("WARNING: Waiting for all daemonsets to be ready failed: %v", err)
+	}
+
+	// Log the version of the server and this client.
+	e2elog.Logf("e2e test version: %s", version.Get().GitVersion)
+
+	dc := c.DiscoveryClient
+
+	serverVersion, serverErr := dc.ServerVersion()
+	if serverErr != nil {
+		e2elog.Logf("Unexpected server error retrieving version: %v", serverErr)
+	}
+	if serverVersion != nil {
+		e2elog.Logf("kube-apiserver version: %s", serverVersion.GitVersion)
+	}
+
+	if TestContext.NodeKiller.Enabled {
+		nodeKiller := NewNodeKiller(TestContext.NodeKiller, c, TestContext.Provider)
+		nodeKillerStopCh = make(chan struct{})
+		go nodeKiller.Run(nodeKillerStopCh)
+	}
+	return nil
+}
+
+// Similar to SynchronizedBeforeSuite, we want to run some operations only once (such as collecting cluster logs).
+// Here, the order of functions is reversed; first, the function which runs everywhere,
+// and then the function that only runs on the first Ginkgo node.
+func CleanupSuite() {
+	// Run on all Ginkgo nodes
+	e2elog.Logf("Running AfterSuite actions on all nodes")
+	RunCleanupActions()
+	if TestContext.NodeKiller.Enabled {
+		close(nodeKillerStopCh)
+	}
+}
+
+// Run a test container to try and contact the Kubernetes api-server from a pod, wait for it
+// to flip to Ready, log its output and delete it.
+func runKubernetesServiceTestContainer(c clientset.Interface, ns string) {
```

Comment:
Not sure if this belongs in this file, seems more like a helper function than anything for initializing the e2e framework. Thoughts about putting this in `test/e2e/framework/util.go` for now?